From 269c21d5f59286009f5d3e22b3c441b57c767b8b Mon Sep 17 00:00:00 2001
From: Macross Chen <macross.chen@mediatek.com>
Date: Mon, 25 Jul 2022 10:47:00 +0800
Subject: [PATCH] gstrawvideoparse: Fix frame size calculation

fix the last_plane_size of GST_VIDEO_FORMAT_NV12_16L32S UV subsample
 case in gst_raw_video_parse_update_info().

Signed-off-by: Macross Chen <macross.chen@mediatek.com>
---
 gst/rawparse/gstrawvideoparse.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/gst/rawparse/gstrawvideoparse.c b/gst/rawparse/gstrawvideoparse.c
index ab0fa65..0e76ef9 100644
--- a/gst/rawparse/gstrawvideoparse.c
+++ b/gst/rawparse/gstrawvideoparse.c
@@ -1163,6 +1163,15 @@ gst_raw_video_parse_update_info (GstRawVideoParseConfig * config)
     gint y_tiles = GST_VIDEO_TILE_Y_TILES (stride);
     gint tile_width = 1 << GST_VIDEO_FORMAT_INFO_TILE_WS (info->finfo);
     gint tile_height = 1 << GST_VIDEO_FORMAT_INFO_TILE_HS (info->finfo);
+
+    if (GST_VIDEO_FORMAT_INFO_FORMAT(info->finfo) == GST_VIDEO_FORMAT_NV12_16L32S) {
+      /* UV is subsampled */
+      /* issue[1340] was submitted to freedesktop for tracking */
+      if (last_plane == 1) {
+        tile_height = tile_height / 2;
+      }
+    }
+
     last_plane_size = x_tiles * y_tiles * tile_width * tile_height;
   } else {
     gint comp[GST_VIDEO_MAX_COMPONENTS];
-- 
2.18.0

